### PR TITLE
[ROCm] Adding ROCDL Dialect.

### PR DESCRIPTION
**Pull request by deven-amd**
This PR introduces the ROCDL Dialect (i.e. the ROCDL ops + the code to lower those ROCDL ops to LLVM intrinsics/functions). Think of ROCDL Dialect as analogous to the NVVM Dialect, but for AMD GPUs. This PR contains just the essentials needed to get a simple example up and running. We expect to make further additions to the ROCDL Dialect.

This is the first of 3 PRs we will be filing. Once this PR is merged it will be followed by
 * a PR to add a pass that lowers GPU Dialect to ROCDL Dialect
 * a PR to add a "mlir-rocm-runner" utility

Note:  This PR is a follow-up on PR #63. It builds on the work done by @whchung in that PR, and addresses most of the review comments in that PR.

---------------------------------

@whchung @joker-eph

### CODE REVIEW

File: lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  return failure(parser.parseOptionalAttributeDict(result.attributes) ||
+                 parser.parseColonType(type) ||
+                 parser.addTypeToList(type, result.types));
+}
+
+//===----------------------------------------------------------------------===//
+// ROCDLDialect initialization, type parsing, and registration.
+//===----------------------------------------------------------------------===//
+
+// TODO(herhut): This should be the llvm.rocdl dialect once this is supported.
```
Comment (joker-eph) [inline lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp]: I see that you went copy-paste all the way, even the TODO is left for Stephan ;)

File: lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      >();
+
+  // Support unknown operations because not all ROCDL operations are registered.
+  allowUnknownOperations();
+}
+
+#define GET_OP_CLASSES
+#include "mlir/Dialect/LLVMIR/ROCDLOps.cpp.inc"
+
+static DialectRegistration<ROCDLDialect> nvvmDialect;
```
Comment (joker-eph) [inline lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp]: Nit: the name should not be `nvvmDialect`

File: lib/Target/LLVMIR/ConvertToROCDLIR.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+namespace {
+static llvm::Value *createIntrinsicCall(llvm::IRBuilder<> &builder,
+                                        llvm::Intrinsic::ID intrinsic,
+                                        ArrayRef<llvm::Value *> args = {}) {
+  llvm::Module *module = builder.GetInsertBlock()->getModule();
+  llvm::Function *fn = llvm::Intrinsic::getDeclaration(module, intrinsic);
+  return builder.CreateCall(fn, args);
+}
+
+static llvm::Value *createDeviceFunctionCall(llvm::IRBuilder<> &builder,
```
Comment (joker-eph) [inline lib/Target/LLVMIR/ConvertToROCDLIR.cpp]: Can you add a comment to document this function?
The name is generic but the signature and the implementation are very coupled to the specific accessors currently defined `__ockl_get_{local, global}_size`
(it only generates call to a function with the provided name that takes a single constant i32, and returns a single i32)

File: include/mlir/Dialect/LLVMIR/ROCDLDialect.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+//
+// This file defines the ROCDL IR dialect in MLIR, containing ROCDL operations
+// and ROCDL specific extensions to the LLVM type system.
```
[the remark this answers is not in the document]
Comment (deven-amd) [reply include/mlir/Dialect/LLVMIR/ROCDLDialect.h]: ROCDL is short for ROCm-Device-Libs.

Unfortunately there is no spec for it. My understanding is that there is no formal IR definition associated with it, and that the ROCDL Dialect will grow to eventually just represent the subset of the LLVM IR extensions supported by the AMDGPU backend 
(https://llvm.org/docs/AMDGPUUsage.html)

@whchung, feel free to add/correct my answer

File: lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp (hunk 1 above)
Comment (deven-amd) [reply lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp]: Yes, this was by design :)

File: lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp (hunk 2 above)
Comment (deven-amd) [reply lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp]: This was not by design...will fix and push out a new commit

File: include/mlir/Dialect/LLVMIR/ROCDLDialect.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (joker-eph) [reply include/mlir/Dialect/LLVMIR/ROCDLDialect.h]: Can we add a link to the LLVM page as a reference then?

File: lib/Target/LLVMIR/ConvertToROCDLIR.cpp (hunk 3 above)
Comment (deven-amd) [reply lib/Target/LLVMIR/ConvertToROCDLIR.cpp]: will do.

We expect this function and its interface to change in the future, as we add support for more stuff. In the meantime, I will add a comment that matches its current usage.

**Comment by joker-eph** [comment]
The test failed on MacOS:
```
FAIL: MLIR :: Dialect/LLVMIR/rocdl.mlir (130 of 293)
******************** TEST 'MLIR :: Dialect/LLVMIR/rocdl.mlir' FAILED ********************
Script:
--
: 'RUN: at line 1';   /tmpfs/src/build/bin/mlir-translate -mlir-to-rocdlir /Volumes/BuildData/tmpfs/src/github/mlir/test/Dialect/LLVMIR/rocdl.mlir | /tmpfs/src/build/bin/FileCheck /Volumes/BuildData/tmpfs/src/github/mlir/test/Dialect/LLVMIR/rocdl.mlir
--
Exit Code: 2

Command Output (stderr):
--
Stack dump:
0.	Program arguments: /tmpfs/src/build/bin/mlir-translate -mlir-to-rocdlir /Volumes/BuildData/tmpfs/src/github/mlir/test/Dialect/LLVMIR/rocdl.mlir 
0  mlir-translate           0x00000001073501a8 llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 40
1  mlir-translate           0x000000010734f208 llvm::sys::RunSignalHandlers() + 248
2  mlir-translate           0x00000001073507c8 SignalHandler(int) + 264
3  libsystem_platform.dylib 0x00007fff5570df5a _sigtramp + 26
4  libsystem_platform.dylib 0x0000000000001218 _sigtramp + 18446603339082707672
5  mlir-translate           0x00000001071ff2ed llvm::CallInst::Create(llvm::FunctionType*, llvm::Value*, llvm::ArrayRef<llvm::Value*>, llvm::ArrayRef<llvm::OperandBundleDefT<llvm::Value*> >, llvm::Twine const&, llvm::Instruction*) + 333
6  mlir-translate           0x00000001071ff042 llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>::CreateCall(llvm::FunctionType*, llvm::Value*, llvm::ArrayRef<llvm::Value*>, llvm::Twine const&, llvm::MDNode*) + 66
7  mlir-translate           0x0000000107200aed (anonymous namespace)::createDeviceFunctionCall(llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&, llvm::StringRef, int) + 189
8  mlir-translate           0x0000000107200105 (anonymous namespace)::ModuleTranslation::convertOperation(mlir::Operation&, llvm::IRBuilder<llvm::ConstantFolder, llvm::IRBuilderDefaultInserter>&) + 853
9  mlir-translate           0x0000000107367ae7 mlir::LLVM::ModuleTranslation::convertBlock(mlir::Block&, bool) + 695
10 mlir-translate           0x0000000107368dc5 mlir::LLVM::ModuleTranslation::convertOneFunction(mlir::FuncOp) + 2005
11 mlir-translate           0x0000000107369470 mlir::LLVM::ModuleTranslation::convertFunctions() + 1536
12 mlir-translate           0x00000001071ffba4 mlir::translateModuleToROCDLIR(mlir::ModuleOp) + 372
13 mlir-translate           0x0000000107200b89 std::__1::__function::__func<$_0, std::__1::allocator<$_0>, mlir::LogicalResult (mlir::ModuleOp, llvm::raw_ostream&)>::operator()(mlir::ModuleOp&&, llvm::raw_ostream&) + 25
14 mlir-translate           0x00000001072a618d std::__1::__function::__func<mlir::TranslationParser::TranslationParser(llvm::cl::Option&)::$_1, std::__1::allocator<mlir::TranslationParser::TranslationParser(llvm::cl::Option&)::$_1>, mlir::LogicalResult (std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer> >, llvm::raw_ostream&, mlir::MLIRContext*)>::operator()(std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer> >&&, llvm::raw_ostream&, mlir::MLIRContext*&&) + 189
15 mlir-translate           0x0000000107250663 main + 1155
16 libdyld.dylib            0x00007fff553ff015 start + 1
FileCheck error: '-' is empty.
FileCheck command line:  /tmpfs/src/build/bin/FileCheck /Volumes/BuildData/tmpfs/src/github/mlir/test/Dialect/LLVMIR/rocdl.mlir

```

File: include/mlir/Dialect/LLVMIR/ROCDLDialect.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (deven-amd) [reply include/mlir/Dialect/LLVMIR/ROCDLDialect.h]: These pages have some more information as well.

https://github.com/RadeonOpenCompute/ROCm-Device-Libs/blob/master/doc/OCML.md
https://github.com/RadeonOpenCompute/ROCm-Device-Libs/blob/master/doc/OCKL.md

I will add all the relevant links in the comments.

**Comment by deven-amd** [comment]
hmm...same failure on Ubuntu as well, but test is passing for me locally

```
[1/1] Running lit suite /home/rocm-user/llvm-project/llvm/projects/mlir/test/Dialect/LLVMIR
Testing Time: 0.22s
  Expected Passes    : 7
```

looking into it further

File: include/mlir/Dialect/LLVMIR/ROCDLDialect.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (deven-amd) [reply include/mlir/Dialect/LLVMIR/ROCDLDialect.h]: pushed out a new commit with the requested comment added

File: lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp (hunk 2 above)
Comment (deven-amd) [reply lib/Dialect/LLVMIR/IR/ROCDLDialect.cpp]: done

File: lib/Target/LLVMIR/ConvertToROCDLIR.cpp (hunk 3 above)
Comment (deven-amd) [reply lib/Target/LLVMIR/ConvertToROCDLIR.cpp]: done

**Comment by deven-amd** [comment]
Still do not know the cause of the CI failures. My local runs pass, even the one I tried on a machine with no ROCm libraries installed!

Hoping the CI failure was an anomaly and the re-run for the new commit passes.

**Comment by deven-amd** [comment]
links to the CI failure logs from the first run
(just in case we need to refer to them again)

Mac : https://source.cloud.google.com/results/invocations/f5be02e6-c7da-4c2e-b72a-0dc7c946b568/log

Ubuntu : https://source.cloud.google.com/results/invocations/0ceb0d9a-d6c2-455a-b2ec-f55181683180/log

**Comment by joker-eph** [comment]
> Hoping the CI failure was an anomaly

I'm not a big believer in the cosmic ray effect here ;)

Did you try the same configuration as the CI? (`Release` mode?)

**Comment by deven-amd** [comment]
Default is `Release` mode..right?

Here are  the commands I am using to do the build and run

```
cmake -G Ninja ../llvm -DLLVM_BUILD_EXAMPLES=ON -DLLVM_ENABLE_CXX1Y=Y -DLLVM_TARGETS_TO_BUILD="host"
cmake --build . --target check-mlir
```

I see that the `-DLLVM_ENABLE_CXX1Y=Y` is no longer present in the instructions provided on README.md...wonder if that is the difference. Let me try that out.

**Comment by deven-amd** [comment]
that is not it either...

```
[1584/1584] Running the MLIR regression tests
Testing Time: 4.08s
  Expected Passes    : 289
  Unsupported Tests  : 4
```

**Comment by deven-amd** [comment]
How do I go about replicating exactly what the Ubuntu CI job is running?

**Comment by deven-amd** [comment]
able to reproduce the error locally now.

It is a `Debug` vs `Release` build issue (default build is `Debug` ... )

looking into it

**Comment by deven-amd** [comment]
found my bug, and fixed it. 

@joker-eph , I have squashed all my changes into one commit, and pushed it out.  Please re-approve.

thanks

deven

**Comment by deven-amd** [comment]
This PR is missing a test file 

* `test/Target/rocdl.mlir` is missing
  * also the contents of `test/Dialect/LLVMIR/rocdl.mlir` contents should be moved here
* `test/Dialect/LLVMIR/rocdl.mlir` need updating (use `mlir-opt` instead of `mlir-transform`)

will push out these changes as a separate commit in the next PR (Lowering GPU Dialect to ROCDL Dialect), which will be filed once this one is merged.

@joker-eph , let me know if you would rather have that change in this PR

**Comment by joker-eph** [comment]
I rather have these in the current PR if you can push?

File: test/Dialect/LLVMIR/rocdl.mlir [hunk 5]
```diff
@@ -0,0 +1,35 @@
+// RUN: mlir-translate -mlir-to-rocdlir %s | FileCheck %s
```
Comment (joker-eph) [inline test/Dialect/LLVMIR/rocdl.mlir]: This folder should test the ops parsing, the verifiers (if any), and custom printing. This test does not seem in the right place, it seems like it is achieving the same thing as: https://github.com/tensorflow/mlir/blob/master/test/Target/nvvmir.mlir

Comment (joker-eph) [reply test/Dialect/LLVMIR/rocdl.mlir]: (Actually this is related to what you called-out, I noticed in parallel)

**Comment by deven-amd** [comment]
> I rather have these in the current PR if you can push?

done.  Please re-approve. thanks.

**Pull request closed without merging**